tripy/docs/pre0_user_guides/02-compiler.md

```diff
@@ -0,0 +1,169 @@
+
+```py
+# doc: no-eval
+inp_info = tp.InputInfo(shape=((1, 8, 16), 2), dtype=tp.float32)
+fast_geglu = compiler.compile(inp_info)
+print(fast_geglu(inp))
+
+# Let's change the shape of input to (2, 2)
+inp = tp.Tensor([[1., 2.], [2., 3.]], dtype=tp.float32)
+print(fast_geglu(inp))
+```
+
+### Errors for out of bounds inference
+If the user provides input that does not comply with the dynamic shape constraint given to the compiler, `Tripy` will produce an error with relevant information:
+
+```py
+# doc: no-eval
+from tests import helper
+
+inp = tp.ones((32, 2), dtype=tp.float32)
+def test_oob(self):
+        with helper.raises(tp.TripyException, match="Unexpected tensor shape."):
+            fast_geglu(inp).eval()
+
+# E       Unexpected tensor shape.
+# E           For parameter `x`, expected the tensor shape `shape(32, 2)` to be within bounds for all dimensions. However, dimension 0 has a shape of 32, which is not within the expected bounds of (1, 16). Note: The provided argument was: 
+# E       
+# E           --> /tripy/tripy/frontend/trace/ops/fill.py:121 in full()
+# E                 |
+# E             121 |     return full_impl(shape, value, dtype, output_rank)
+# E                 | 
+# E       
+# E           --> /tripy/tripy/frontend/ops/tensor_initializers.py:57 in ones()
+# E                 |
+# E              57 |     return full(shape, 1, dtype)
+# E                 |            ^^^^^^^^^^^^^^^^^^^^^ --- required from here
+# E       
+```
+
+### Serializing the executable to disk
+
+A compiled executable can be serialized to disk and then used for deployment.
+
+Saving an executable to disk:
+
+```py
+# doc: no-eval
+import tempfile, os
+temp_dir = tempfile.mkdtemp()
+executable_file_path = os.path.join(temp_dir, "executable.json")
+fast_geglu.save(executable_file_path)
+```
+
+Reading an executable and running inference:
+
+```py
+# doc: no-eval
+inp = tp.Tensor([[1., 2.], [2., 3.]], dtype=tp.float32)
+loaded_fast_geglu = tp.Executable.load(executable_file_path)
+print(loaded_fast_geglu(inp))
+os.remove(executable_file_path)
+```
+
+### Querying properties of the executable
+
+You can also query properties about the executable:
+
+```py
+# doc: no-eval
+loaded_fast_geglu.get_input_info()
+# [ArgInfo(shape_bounds=((1, 16), (2, 2)), dtype=float32)]
+
+loaded_fast_geglu.get_output_info()
+# [ArgInfo(shape_bounds=((1, 16), (8, 8)), dtype=float32)]
+```
+
+## Expectations for pure functions
+
+Tripy expects that the functions that will be compiled are pure functions with no side effects. Consider this example:
+
+```py
+# doc: no-eval
+def add_times_two(a, b):
+    c = a + b
+    print(c)
```

Comment:
I'm trying to think of ways to make this better - one idea I had was to mark tensors as being for tracing only. Any operations that take any tracing-only inputs will generate all tracing-only outputs (we can maybe do that in the build function). In` __repr__` we can emit an error if the tensor is supposed to be tracing only

Reply:
@pmarathe25  I am resolving this comment to merge the MR, will file an issue to discuss this further.